cnf/ext/central.mvn

```diff
@@ -37,6 +37,7 @@ commons-codec:commons-codec:1.10
 
 org.apache.felix:org.apache.felix.configadmin:1.2.4
 org.apache.felix:org.apache.felix.log:1.2.0
+org.apache.felix:org.apache.felix.framework:3.2.2
```

Comment:
Since 3.2.2 is the implementation of OSGi 4.2 whereas the previously added Felix 4.2.1 is the implementation of OSGi 4.3

Reply:
Perhaps, but we do not need to test on an old framework. We can test on a new framework since old bundles will run on new framework. So please use the newest Felix for tests.

Reply:
I also thought the same that it should work actually in newer versions of Felix too but the tests fail with the following stacktrace:

```
java.lang.NoSuchMethodError: org.osgi.framework.BundleEvent.<init>(ILorg/osgi/framework/Bundle;Lorg/osgi/framework/Bundle;)V
	at org.apache.felix.framework.Felix.fireBundleEvent(Felix.java:4584)
	at org.apache.felix.framework.Felix.installBundle(Felix.java:3199)
	at org.apache.felix.framework.BundleContextImpl.installBundle(BundleContextImpl.java:167)
	at aQute.bnd.remote.junit.JUnitFramework.testbundle(JUnitFramework.java:705)
	at aQute.bnd.remote.junit.JUnitFramework.start(JUnitFramework.java:664)
	at aQute.bnd.remote.junit.JUnitFrameworkBuilder.create(JUnitFrameworkBuilder.java:165)
	at aQute.bnd.remote.junit.HelloTest.testActivator(HelloTest.java:57)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:2
```

Reply:
This must be because you have core-4.2 on the -runpath. The framework impl must supply the core api.

Reply:
The core jar is only to be used for compilation. It should never be on the classpath when running a framework.

Reply:
Agreed but how did the the API jar end up in `-runpath`? It is only specified in `-buildpath`.

Reply:
My guess is https://github.com/bndtools/bndtools/issues/1922 which is fixed in the current 4.2 bndtools builds. Are you using the latest 4.2 bndtools? You may find the test passes from the gradle build since it did not have the above issue.

Reply:
I am using the recent snapshot version but the gradle build is still failing. I am not sure if it is the transitive classpath calculation problem. The `System.getProperty("java.class.path")` also returns the following: 

```
System.getProperty("java.class.path")
	 (java.lang.String) /Users/amitmondal/git/bnd/biz.aQute.bnd.remote.junit/bin:/Users/amitmondal/git/bnd/biz.aQute.bnd.remote.junit/bin_test:/Users/amitmondal/.m2/repository/org/osgi/osgi.annotation/7.0.0/osgi.annotation-7.0.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.core/4.2.0/org.osgi.core-4.2.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.service.component/1.3.0/org.osgi.service.component-1.3.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.service.component.annotations/1.3.0/org.osgi.service.component.annotations-1.3.0.jar:/Users/amitmondal/git/bnd/aQute.libg/bin:/Users/amitmondal/git/bnd/aQute.libg/bin_test:/Users/amitmondal/.m2/repository/org/slf4j/slf4j-api/1.7.25/slf4j-api-1.7.25.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.util.function/1.1.0/org.osgi.util.function-1.1.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.util.promise/1.1.1-SNAPSHOT/org.osgi.util.promise-1.1.1-SNAPSHOT.jar:/Users/amitmondal/.m2/repository/org/apache/servicemix/bundles/org.apache.servicemix.bundles.junit/4.12_1/org.apache.servicemix.bundles.junit-4.12_1.jar:/Users/amitmondal/.m2/repository/org/assertj/assertj-core/3.10.0/assertj-core-3.10.0.jar:/Users/amitmondal/.m2/repository/org/osgi/osgi.core/6.0.0/osgi.core-6.0.0.jar:/Users/amitmondal/.m2/repository/org/slf4j/slf4j-simple/1.7.25/slf4j-simple-1.7.25.jar:/Users/amitmondal/git/bnd/biz.aQute.bndlib/bin:/Users/amitmondal/git/bnd/biz.aQute.bndlib/bin_test:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.namespace.contract/1.0.0/org.osgi.namespace.contract-1.0.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.namespace.extender/1.0.1/org.osgi.namespace.extender-1.0.1.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.namespace.implementation/1.0.0/org.osgi.namespace.implementation-1.0.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.namespace.service/1.0.0/org.osgi.namespace.service-1.0.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.service.log/1.3.0/org.osgi.service.log-1.3.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.service.repository/1.1.0/org.osgi.service.repository-1.1.0.jar:/Users/amitmondal/git/bnd/biz.aQute.bnd.annotation/bin:/Users/amitmondal/git/bnd/biz.aQute.bnd.annotation/bin_test:/Users/amitmondal/git/bnd/biz.aQute.bndlib/generated/biz.aQute.bndlib.jar:/Users/amitmondal/git/bnd/biz.aQute.repository/bin:/Users/amitmondal/git/bnd/biz.aQute.repository/bin_test:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.service.coordinator/1.0.2/org.osgi.service.coordinator-1.0.2.jar:/Users/amitmondal/.m2/repository/org/tukaani/xz/1.5/xz-1.5.jar:/Users/amitmondal/.m2/repository/org/mockito/mockito-core/2.23.0/mockito-core-2.23.0.jar:/Users/amitmondal/.m2/repository/net/bytebuddy/byte-buddy/1.9.0/byte-buddy-1.9.0.jar:/Users/amitmondal/.m2/repository/org/objenesis/objenesis/2.6/objenesis-2.6.jar:/Users/amitmondal/.m2/repository/org/apache/geronimo/specs/geronimo-servlet_2.5_spec/1.2/geronimo-servlet_2.5_spec-1.2.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-continuation/7.6.3.v20120416/jetty-continuation-7.6.3.v20120416.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-http/7.6.3.v20120416/jetty-http-7.6.3.v20120416.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-io/7.6.3.v20120416/jetty-io-7.6.3.v20120416.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-security/7.6.3.v20120416/jetty-security-7.6.3.v20120416.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-server/7.6.3.v20120416/jetty-server-7.6.3.v20120416.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-util/7.6.3.v20120416/jetty-util-7.6.3.v20120416.jar:/Users/amitmondal/git/bnd/biz.aQute.http.testservers/bin:/Users/amitmondal/git/bnd/biz.aQute.http.testservers/bin_test:/Users/amitmondal/.m2/repository/org/nanohttpd/nanohttpd/2.2.0/nanohttpd-2.2.0.jar:/Users/amitmondal/.m2/repository/org/bouncycastle/bcpkix-jdk15on/1.54/bcpkix-jdk15on-1.54.jar:/Users/amitmondal/.m2/repository/org/bouncycastle/bcprov-jdk15on/1.54/bcprov-jdk15on-1.54.jar:/Users/amitmondal/.m2/repository/org/littleshoot/littleproxy/1.1.2/littleproxy-1.1.2.jar:/Users/amitmondal/.m2/repository/com/google/guava/guava/19.0/guava-19.0.jar:/Users/amitmondal/.m2/repository/org/apache/commons/commons-lang3/3.4/commons-lang3-3.4.jar:/Users/amitmondal/.m2/repository/commons-codec/commons-codec/1.10/commons-codec-1.10.jar:/Users/amitmondal/git/bnd/biz.aQute.http.testservers/generated/biz.aQute.http.testservers.jar:/Users/amitmondal/git/bnd/biz.aQute.repository/generated/biz.aQute.repository.jar:/Users/amitmondal/java-latest/Eclipse.app/Contents/Eclipse/configuration/org.eclipse.osgi/225/0/.cp:/Users/amitmondal/java-latest/Eclipse.app/Contents/Eclipse/configuration/org.eclipse.osgi/224/0/.cp:/Users/amitmondal/java-latest/Eclipse.app/Contents/Eclipse/configuration/org.eclipse.osgi/227/0/.cp/lib/javaagent-shaded.jar
```

It comprises both OSGi Core 4.2 and OSGi Core 6.0 which seems really strange.

Reply:
This may be a defect in the design of JunitFramework in that it is too connected to the build environment of the test project. I don't have any real experience using JunitFramework but it seems to me that there ought to be a disconnect between the class path of the framework being tested and the class path of the test case. It seems that the test case class path bleeds into the framework being tested which mean you have to have an exact match of the core version in test case class path and the framework being tested. While this could be matched (e.g. use a 4.2 framework version here), it is entirely permissible for framework implementations to customize the implementation of the core classes and so the framework must be able to use its own core classes rather than the vanilla ones used to build the test case. 

I think you need to take this up with @pkriens.